### PR TITLE
Introduce `QobjEvo` and use `SciMLOperators` for time evolution

### PR DESCRIPTION
**Pull request by albertomercurio**
## Description
With this PR I remove the obsolete operators like `OperatorSum` and `TimeDependentOperatorSum`, by introducing the `QobjEvo` operator, like in python. This `struct` takes advantage of the [SciMLOperators.jl](https://github.com/SciML/SciMLOperators.jl) ecosystem, in order to handle in a lazy way complex operators like sum of them and time-dependent ones.

The `ObjEvo` is and `AbstractQuantumObject`, with the same fields as the `QuantumObject`, so they share many of the methods.

The introduction of using [SciMLOperators.jl](https://github.com/SciML/SciMLOperators.jl) would push us towards the support of automatic differentiation of `sesolve`, `mesolve` and `mcsolve`, allowing the possibility to make quantum optimal control very efficiently.

I made [this PR](https://github.com/SciML/SciMLOperators.jl/pull/249) to improve the performance of complex `SciMLOperators` like our sum of time-dependent operators. So, we have to wait for a new release of the package first.

Todo list:
- [x] Introduce `QobjEvo` operator
- [x] Remove `tlist` from the parameters, and add it directly as argument
- [x] Make `TimeEvolutionSol` type-stable
- [x] replace `t_l` with a more readable `tlist`
- [x] Update docstrings
- [x] Remove `OperatorSum` and `TimeDependentOperatorSum`
- [x] Update `sesolve` with the support to `QobjEvo`
- [x] Update `mesolve` with the support to `QobjEvo`
- [x] Update `mcsolve` with the support to `QobjEvo`
- [x] Update `ssesolve` with support to `QobjEvo`
- [x] Remove `alg` as argument for all the Problem functions
- [x] Add tests for all the added methods
- [x] Wait for merging of [this PR](https://github.com/SciML/SciMLOperators.jl/pull/249) and new release of [SciMLOperators.jl](https://github.com/SciML/SciMLOperators.jl) 

I will add the documentation in the next PRs. This one is already heavy.

### CODE REVIEW

File: src/qobj/functions.jl [hunk 1]
```diff
@@ -191,13 +189,17 @@ Quantum Object:   type=Operator   dims=[20, 20]   size=(400, 400)   ishermitian=
...
-LinearAlgebra.kron(A::QuantumObject) = A
-function LinearAlgebra.kron(A::Vector{<:QuantumObject})
+function LinearAlgebra.kron(
+    A::AbstractQuantumObject{DT1,OpType},
+    B::AbstractQuantumObject{DT2,OpType},
+) where {DT1,DT2,OpType<:Union{KetQuantumObject,BraQuantumObject,OperatorQuantumObject}}
+    if A isa QuantumObjectEvolution || B isa QuantumObjectEvolution
+        return QuantumObjectEvolution(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
+    end
+    return QuantumObject(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
```
Comment (ytdHuang) [inline src/qobj/functions.jl:199]: I think we can extend the basic julia function `promote_type` (maybe put it in the file `utilities.jl` ?) for this.
something like
```julia
promote_type(A::QuantumObjectEvolution, B::QuantumObjectEvolution) = get_typename_wrapper(A)
promote_type(A::QuantumObjectEvolution, B::QuantumObject) = get_typename_wrapper(A)
promote_type(A::QuantumObject, B::QuantumObjectEvolution) = get_typename_wrapper(B)
promote_type(A::QuantumObject, B::QuantumObject) = get_typename_wrapper(A)
```

and therefore, this kind of `if`-condition block can be simplified as
```julia
QType = promote_type(A, B)
QType(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
```

File: src/progress_bar.jl [hunk 2]
```diff
@@ -1,6 +1,6 @@
 export ProgressBar, next!
 
-struct ProgressBar{CT,T1<:Integer,T2<:Real,T3,T4<:Real,LT}
+mutable struct ProgressBar{CT,T1<:Integer,T2<:Real,T3,T4<:Real,LT}
```
Comment (ytdHuang) [inline src/progress_bar.jl:3]: This change should be in another PR.
maybe you need to merge from main branch locally first.

File: src/qobj/arithmetic_and_attributes.jl [hunk 3]
```diff
@@ -36,84 +36,81 @@ end
...
-            B::QuantumObject{<:AbstractArray{T2},OpType},
-        ) where {T1,T2,OpType<:QuantumObjectType}
-            A.dims != B.dims &&
-                throw(DimensionMismatch("The two quantum objects are not of the same Hilbert dimension."))
+        function LinearAlgebra.$op(A::AbstractQuantumObject, B::AbstractQuantumObject)
+            check_dims(A, B)
+            if A isa QuantumObjectEvolution || B isa QuantumObjectEvolution
+                return QuantumObjectEvolution($(op)(A.data, B.data), A.type, A.dims)
+            end
             return QuantumObject($(op)(A.data, B.data), A.type, A.dims)
```
Comment (ytdHuang) [inline src/qobj/arithmetic_and_attributes.jl:44]: I think we can extend the basic julia function `promote_type` (maybe put it in the file `utilities.jl` ?) for this.
something like
```julia
promote_type(A::QuantumObjectEvolution, B::QuantumObjectEvolution) = get_typename_wrapper(A)
promote_type(A::QuantumObjectEvolution, B::QuantumObject) = get_typename_wrapper(A)
promote_type(A::QuantumObject, B::QuantumObjectEvolution) = get_typename_wrapper(B)
promote_type(A::QuantumObject, B::QuantumObject) = get_typename_wrapper(A)
```

and therefore, this kind of `if`-condition block can be simplified as
```julia
QType = promote_type(A, B)
QType($(op)(A.data, B.data), A.type, A.dims)
```

File: src/qobj/arithmetic_and_attributes.jl [hunk 4]
```diff
@@ -125,93 +122,91 @@ Note that `A` and `B` should be [`Ket`](@ref) or [`OperatorKet`](@ref)
...
     A.dims != B.dims && throw(DimensionMismatch("The quantum objects are not of the same Hilbert dimension."))
     return LinearAlgebra.dot(A.data, B.data)
 end
 
 @doc raw"""
-    dot(i::QuantumObject, A::QuantumObject j::QuantumObject)
+    dot(i::QuantumObject, A::AbstractQuantumObject j::QuantumObject)
 
-Compute the generalized dot product `dot(i, A*j)` between three [`QuantumObject`](@ref): ``\langle i | \hat{A} | j \rangle``
+Compute the generalized dot product `dot(i, A*j)` between three [`AbstractQuantumObject`](@ref): ``\langle i | \hat{A} | j \rangle``
```
Comment (ytdHuang) [inline src/qobj/arithmetic_and_attributes.jl:135]: maybe change this as
```julia
Compute the generalized dot product `dot(i, A*j)` between [`AbstractQuantumObject`](@ref) `A` and two [`QuantumObject`](@ref) (`i` and `j`), namely ``\langle i | \hat{A} | j \rangle``
```

File: src/qobj/boolean_functions.jl [hunk 5]
```diff
@@ -6,71 +6,79 @@ export isket, isbra, isoper, isoperbra, isoperket, issuper
 export isunitary
```
Comment (ytdHuang) [inline src/qobj/boolean_functions.jl:6]: Should we also add a new function `isconstant` in the end of first line ?
I saw that QuTiP also got this function

File: src/qobj/functions.jl (hunk 1 above)
Comment (ytdHuang) [inline src/qobj/functions.jl:199]: can also use `promote_type` mentioned above

File: src/qobj/quantum_object.jl [hunk 6]
```diff
@@ -226,6 +232,11 @@ _check_dims(dims::Any) = throw(
     ),
 )
 
+function check_dims(A::AbstractQuantumObject, B::AbstractQuantumObject)
+    A.dims != B.dims && throw(DimensionMismatch("The two quantum objects don't have the same Hilbert dimension."))
+    return nothing
+end
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:238]: how about moving this function into `utilities.jl`, since it would be used by both `quantum_object.jl` and `quantum_object_evo.jl`

File: src/qobj/quantum_object.jl [hunk 7]
```diff
@@ -369,6 +380,8 @@ SparseArrays.SparseMatrixCSC(A::QuantumObject{<:AbstractMatrix}) =
 SparseArrays.SparseMatrixCSC{T}(A::QuantumObject{<:SparseMatrixCSC}) where {T<:Number} =
     QuantumObject(SparseMatrixCSC{T}(A.data), A.type, A.dims)
 
+get_typename_wrapper(A::AbstractQuantumObject) = Base.typename(typeof(A)).wrapper
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:383]: how about moving this function into `utilities.jl`, since it would be used by both `quantum_object.jl` and `quantum_object_evo.jl`

File: src/qobj/quantum_object.jl [hunk 8]
```diff
@@ -369,6 +380,8 @@ SparseArrays.SparseMatrixCSC(A::QuantumObject{<:AbstractMatrix}) =
...
+
 # functions for getting Float or Complex element type
 _FType(::QuantumObject{<:AbstractArray{T}}) where {T<:Number} = _FType(T)
 _CType(::QuantumObject{<:AbstractArray{T}}) where {T<:Number} = _CType(T)
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:387]: we can change these to
```julia
_FType(::AbstractQuantumObject) = _FType(eltype(T))
_CType(::AbstractQuantumObject) = _CType(eltype(T))
```

File: src/time_evolution/mcsolve.jl [hunk 9]
```diff
@@ -23,6 +23,7 @@ end
 function LindbladJumpAffect!(integrator)
     internal_params = integrator.p
     c_ops = internal_params.c_ops
+    c_ops_herm = internal_params.c_ops_herm
```
Comment (ytdHuang) [inline src/time_evolution/mcsolve.jl:26]: I thought this is done by another PR, maybe you need to merge from main branch locally first.

File: src/time_evolution/sesolve.jl [hunk 10]
```diff
@@ -16,42 +16,34 @@ function _save_func_sesolve(integrator)
...
-        H_t::Union{Nothing,Function,TimeDependentOperatorSum}=nothing,
-        params::NamedTuple=NamedTuple(),
-        progress_bar::Union{Val,Bool}=Val(true),
+    sesolveProblem(H,
+        ψ0,
+        tlist;
+        alg=Tsit5()
+        e_ops = nothing,
+        params=NamedTuple(),
+        progress_bar=Val(true),
```
Comment (ytdHuang) [inline src/time_evolution/sesolve.jl:46]: Not necessary but I think it would be better to add blanks on both sides of `=`

File: src/qobj/quantum_object_evo.jl [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+julia> op1(p, 0.1)
+Quantum Object:   type=Operator   dims=[10, 2]   size=(20, 20)   ishermitian=false
+20×20 SparseMatrixCSC{ComplexF64, Int64} with 28 stored entries:
+⎡⠂⡑⢄⠀⠀⠀⠀⠀⠀⠀⎤
+⎢⠀⠀⠂⡑⢄⠀⠀⠀⠀⠀⎥
+⎢⠀⠀⠀⠀⠂⡑⢄⠀⠀⠀⎥
+⎢⠀⠀⠀⠀⠀⠀⠂⡑⢄⠀⎥
+⎣⠀⠀⠀⠀⠀⠀⠀⠀⠂⡑⎦
+"""
+struct QuantumObjectEvolution{DT<:AbstractSciMLOperator,ObjType<:QuantumObjectType,N} <:
```
Comment (ytdHuang) [inline src/qobj/quantum_object_evo.jl:86]: Do we need to restrict `ObjType<:Union{OperatorQuantumObject,SuperOperatorQuantumObject}` here ?

File: src/progress_bar.jl (hunk 2 above)
Comment (albertomercurio) [reply src/progress_bar.jl:3]: I have just done "merge from upstream:main", but now it seems a like a change

File: src/qobj/boolean_functions.jl [hunk 12]
```diff
@@ -81,3 +89,10 @@ Note that all the keyword arguments will be passed to `Base.isapprox`.
 """
 isunitary(U::QuantumObject{<:AbstractArray{T}}; kwargs...) where {T} =
     isoper(U) ? isapprox(U.data * U.data', I(size(U, 1)); kwargs...) : false
+
+@doc raw"""
+    isconstant(A::AbstractQuantumObject)
+
+Test whether the [`AbstractQuantumObject`](@ref) `A` is constant in time. For a [`QuantumObject`](@ref), this function returns `false`, while for a [`QuantumObjectEvolution`](@ref), this function returns `true` if the operator is contant in time.
```
Comment (ytdHuang) [inline src/qobj/boolean_functions.jl:96]: Typo here, 
```
For a [`QuantumObject`](@ref), this function always returns `true`
```

Comment (ytdHuang) [reply src/qobj/boolean_functions.jl:96]: I checked that
```julia
isconstant([1 2; 3 4]) # true
```

File: src/qobj/quantum_object.jl [hunk 13]
```diff
@@ -7,108 +7,7 @@ Also support for fundamental functions in Julia standard library:
     - SparseArrays: sparse, nnz, nonzeros, rowvals, droptol!, dropzeros, dropzeros!, SparseVector, SparseMatrixCSC
 =#
 
-export AbstractQuantumObject, QuantumObject
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:10]: we should also modify the above multi-row comments

File: src/time_evolution/sesolve.jl [hunk 14]
```diff
@@ -184,38 +169,27 @@ Time evolution of a closed quantum system using the Schrödinger equation:
...
 
-function sesolve(prob::ODEProblem, alg::OrdinaryDiffEqAlgorithm = Tsit5())
+function sesolve(prob::ODEProblem, tlist::AbstractVector, alg::OrdinaryDiffEqAlgorithm = Tsit5())
     sol = solve(prob, alg)
 
     ψt = map(ϕ -> QuantumObject(ϕ, type = Ket, dims = sol.prob.p.Hdims), sol.u)
 
     return TimeEvolutionSol(
-        sol.prob.p.times,
+        tlist,
```
Comment (ytdHuang) [inline src/time_evolution/sesolve.jl:192]: Why are we changing this implementation.
this could lead to ambiguity?
cause the `tlist` here might not be the one used in `ODEProblem`

File: src/qobj/quantum_object.jl [hunk 15]
```diff
@@ -1,114 +1,10 @@
 #=
-This file defines:
-    1. the QuantumObject (Qobj) structure
-    2. all the type structures for QuantumObject
-Also support for fundamental functions in Julia standard library:
-    - Base: show, length, size, eltype, getindex, setindex!, isequal, :(==), isapprox, Vector, Matrix
+This file defines the QuantumObject (Qobj) structure.
+It also implements the fundamental functions in Julia standard library:
     - SparseArrays: sparse, nnz, nonzeros, rowvals, droptol!, dropzeros, dropzeros!, SparseVector, SparseMatrixCSC
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:4]: Missing
```
- Base: real, imag, Vector, Matrix
```

File: src/qobj/quantum_object.jl [hunk 16]
```diff
@@ -294,7 +190,7 @@ function Base.show(
     return show(io, MIME("text/plain"), op_data)
 end
 
-function Base.show(io::IO, QO::QuantumObject{<:AbstractArray{T},OpType}) where {T,OpType<:OperatorQuantumObject}
+function Base.show(io::IO, QO::AbstractQuantumObject)
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:193]: Regarding to `show`
This is the output of `QobjEvo` from python qutip:
```
QobjEvo: dims = [[2], [2]], shape = (2, 2), type = oper, superrep = None, isconstant = False, num_elements = 3
```

Should we separate the method of `show` for `Qobj` and `QobjEvo` ?
That is, rebase `Qobj`'s `show` definition as
```julia
Base.show(io::IO, QO::QuantumObject{<:AbstractArray{T},OpType}) where {T,OpType<:OperatorQuantumObject}
```
and maybe print something like the following for `QobjEvo`:
```
Quantum Object Evolution:   type=Operator,   dims=[2, 2],   size=(2, 2),   ishermitian=true,   isconstant=false, 
  num_elements = 3
# data print of SciMLOperator
```
For the `num_elements`, I only know that we can access `ops` from `AddedOperator` and get the `length` of it, namely
```julia
H::AddedOperator # or TensorProductOperator
length(H.ops)
```

But I'm not sure how to get it from pure `MatrixOperator`.
Or maybe we need to define a new function (maybe call it `_get_QobjEvo_num_element`) and multiple dispatch for different `AbstractSciMLOperator`.

File: src/qobj/quantum_object_base.jl [hunk 17]
```diff
@@ -0,0 +1,160 @@
+#=
+This file defines the AbstractQuantumObject structure, all the type structures for QuantumObject and fundamental functions in Julia standard library.
+Also support for fundamental functions in Julia standard library:
+    - Base: show, length, size, eltype, getindex, setindex!, isequal, :(==), isapprox, Vector, Matrix
+=#
```
Comment (ytdHuang) [inline src/qobj/quantum_object_base.jl:5]: maybe just
```
#=
This file defines the AbstractQuantumObject structure, all the type structures for AbstractQuantumObject, and fundamental functions in Julia standard library:
    - Base: show, length, size, eltype, getindex, setindex!, isequal, :(==), isapprox
=#
```

File: src/steadystate.jl [hunk 18]
```diff
@@ -212,48 +212,34 @@ or
...
-        abstol = abstol,
-        kwargs...,
-    )
+    cb = TerminateSteadyState(abstol, reltol, _steadystate_ode_condition)
+    sol = mesolve(H, ψ0, [ftype(0), ftype(tmax)], c_ops, progress_bar = Val(false), callback = cb)
 
-    ρss = reshape(sol.u[end], N, N)
-    return QuantumObject(ρss, Operator, H.dims)
+    ρss = sol.states[end]
+    return ρss
```
Comment (ytdHuang) [inline src/steadystate.jl:242]: Is it necessary to go through `mesolve` routine ?
In this way, `sol.states` will save both the initial state and final state.

File: src/time_evolution/mesolve.jl [hunk 19]
```diff
@@ -98,71 +103,61 @@ where
 - `prob::ODEProblem`: The ODEProblem for the master equation time evolution.
 """
 function mesolveProblem(
-    H::QuantumObject{MT1,HOpType},
-    ψ0::QuantumObject{<:AbstractArray{T2},StateOpType},
+    H::Union{AbstractQuantumObject{DT1,HOpType},Tuple},
+    ψ0::QuantumObject{DT2,StateOpType},
     tlist,
     c_ops::Union{Nothing,AbstractVector,Tuple} = nothing;
     alg::OrdinaryDiffEqAlgorithm = Tsit5(),
```
Comment (ytdHuang) [inline src/time_evolution/mesolve.jl:110]: We also don't need `alg` for generating `mesolveProblem` right?

Comment (ytdHuang) [reply src/time_evolution/mesolve.jl:110]: maybe also need to check for `sesolve`, `mcsolve`, and `ssesolve`.

File: src/time_evolution/sesolve.jl [hunk 20]
```diff
@@ -62,7 +54,7 @@ Generates the ODEProblem for the Schrödinger time evolution of a quantum system
 
 # Arguments
 
-- `H::QuantumObject`: The Hamiltonian of the system ``\hat{H}``.
+- `H::Union{QuantumObject,Tuple}`: The Hamiltonian of the system ``\hat{H}``.
 - `ψ0::QuantumObject`: The initial state of the system ``|\psi(0)\rangle``.
 - `tlist::AbstractVector`: The time list of the evolution.
 - `alg::OrdinaryDiffEqAlgorithm`: The algorithm used for the time evolution.
```
Comment (ytdHuang) [inline src/time_evolution/sesolve.jl:60]: need to remove this line

File: src/qobj/quantum_object_base.jl (hunk 17 above)
Comment (ytdHuang) [reply src/qobj/quantum_object_base.jl:5]: Note that I have removed `Vector` and `Matrix` since they are not in this file

File: src/steadystate.jl (hunk 18 above)
Comment (albertomercurio) [reply src/steadystate.jl:242]: We should avoid redundancies in the package. If there is something already implemented, we should use it. I can add a `saveat` kwarg (also in the previous implementation it was absent). The actual code called is simpler that what one would think. Thanks to multiple dispatch, the ProgressBar-e_ops callback is not even created, and `e_ops` is just an empty `Tuple` in the latest implementation. So I think that the compiled code is very similar to doing everything manually as before.

File: src/time_evolution/mesolve.jl (hunk 19 above)
Comment (albertomercurio) [reply src/time_evolution/mesolve.jl:110]: Yes, we can remove it.

File: src/qobj/quantum_object.jl (hunk 16 above)
Comment (albertomercurio) [reply src/qobj/quantum_object.jl:193]: For the moment I would leave it as it is. Maybe in a next PR we can do this.

File: src/steadystate.jl (hunk 18 above)
Comment (ytdHuang) [reply src/steadystate.jl]: ok

File: src/time_evolution/mcsolve.jl [hunk 21]
```diff
@@ -117,12 +117,24 @@ function _mcsolve_generate_statistics(sol, i, states, expvals_all, jump_times, j
...
+end
+
 @doc raw"""
     mcsolveProblem(H::QuantumObject{<:AbstractArray{T1},OperatorQuantumObject},
         ψ0::QuantumObject{<:AbstractArray{T2},KetQuantumObject},
         tlist::AbstractVector,
         c_ops::Union{Nothing,AbstractVector,Tuple}=nothing;
-        alg::OrdinaryDiffEqAlgorithm=Tsit5(),
         e_ops::Union{Nothing,AbstractVector,Tuple}=nothing,
         H_t::Union{Nothing,Function,TimeDependentOperatorSum}=nothing,
```
Comment (ytdHuang) [inline src/time_evolution/mcsolve.jl:139]: We also need to remove `H_t` in all docstrings of time evolution solves

File: test/core-test/quantum_objects_evo.jl [hunk 22]
```diff
@@ -81,7 +81,7 @@
         @test adjoint(a2).data == adjoint(a2.data)
 
         N = 10
-        a = QobjEvo(destroy(N), Operator, N)
+        a = QobjEvo(MatrixOperator(sprand(ComplexF64, N, N, 5 / N)), Operator, N)
```
Comment (ytdHuang) [inline test/core-test/quantum_objects_evo.jl:84]: Why do we need to construct `MatrixOperator` byourselves?
I thought we can use `QobjEvo(destroy(N), Operator, N)` directly

File: test/runtests.jl [hunk 23]
```diff
@@ -3,6 +3,7 @@ using Pkg
 using QuantumToolbox
 using QuantumToolbox: position, momentum
 using Random
+using SciMLOperators
```
Comment (ytdHuang) [inline test/runtests.jl:6]: If we don't need to use `MatrixOperator` above, then I think we can remove this.

**Review by ytdHuang — APPROVED** [review]

File: src/qobj/quantum_object_base.jl [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+
+Abstract type for all quantum objects like [`QuantumObject`](@ref) and [`QuantumObjectEvolution`](@ref).
+
+# Example
+```
+julia> sigmax() isa AbstractQuantumObject
+true
+```
+"""
+abstract type AbstractQuantumObject{DataType,ObjType,N} end
```
Comment (ytdHuang) [inline src/qobj/quantum_object_base.jl:27]: We should remove the parameter `N`, namely
```julia
abstract type AbstractQuantumObject{DataType,ObjType} end
```

File: src/qobj/boolean_functions.jl [hunk 25]
```diff
@@ -3,74 +3,82 @@ All boolean functions for checking the data or type in `QuantumObject`
...
-    isoper(A::QuantumObject)
+    isoper(A)
 
-Checks if the [`QuantumObject`](@ref) `A` is a [`OperatorQuantumObject`](@ref).
+Checks if the [`AbstractQuantumObject`](@ref) `A` is a [`OperatorQuantumObject`](@ref). Default case returns `false` for any other inputs.
 """
-isoper(A::QuantumObject{<:AbstractArray{T},OpType}) where {T,OpType<:QuantumObjectType} =
-    OpType <: OperatorQuantumObject
+isoper(A::AbstractQuantumObject) = isoper(typeof(A))
+isoper(::Type{<:AbstractQuantumObject{DT,OperatorQuantumObject,N}}) where {DT,N} = true
```
Comment (ytdHuang) [inline src/qobj/boolean_functions.jl:32]: same here, remove `N`

File: src/qobj/boolean_functions.jl [hunk 26]
```diff
@@ -3,74 +3,82 @@ All boolean functions for checking the data or type in `QuantumObject`
...
-    issuper(A::QuantumObject)
+    issuper(A)
 
-Checks if the [`QuantumObject`](@ref) `A` is a [`SuperOperatorQuantumObject`](@ref).
+Checks if the [`AbstractQuantumObject`](@ref) `A` is a [`SuperOperatorQuantumObject`](@ref). Default case returns `false` for any other inputs.
 """
-issuper(A::QuantumObject{<:AbstractArray{T},OpType}) where {T,OpType<:QuantumObjectType} =
-    OpType <: SuperOperatorQuantumObject
+issuper(A::AbstractQuantumObject) = issuper(typeof(A))
+issuper(::Type{<:AbstractQuantumObject{DT,SuperOperatorQuantumObject,N}}) where {DT,N} = true
```
Comment (ytdHuang) [inline src/qobj/boolean_functions.jl:59]: same here, remove `N`

File: src/qobj/quantum_object.jl [hunk 27]
```diff
@@ -135,7 +32,7 @@ julia> a isa QuantumObject
 true
 ```
 """
-struct QuantumObject{MT<:AbstractArray,ObjType<:QuantumObjectType,N} <: AbstractQuantumObject
+struct QuantumObject{MT<:AbstractArray,ObjType<:QuantumObjectType,N} <: AbstractQuantumObject{MT,ObjType,N}
```
Comment (ytdHuang) [inline src/qobj/quantum_object.jl:35]: same here, remove `N`

File: src/qobj/quantum_object_evo.jl [hunk 28]
```diff
@@ -0,0 +1,418 @@
...
+⎢⠀⠀⠀⠀⠂⡑⢄⠀⠀⠀⎥
+⎢⠀⠀⠀⠀⠀⠀⠂⡑⢄⠀⎥
+⎣⠀⠀⠀⠀⠀⠀⠀⠀⠂⡑⎦
+```
+"""
+struct QuantumObjectEvolution{
+    DT<:AbstractSciMLOperator,
+    ObjType<:Union{OperatorQuantumObject,SuperOperatorQuantumObject},
+    N,
+} <: AbstractQuantumObject{DT,ObjType,N}
```
Comment (ytdHuang) [inline src/qobj/quantum_object_evo.jl:91]: same here, remove `N`

File: test/core-test/quantum_objects_evo.jl (hunk 22 above)
Comment (albertomercurio) [reply test/core-test/quantum_objects_evo.jl:84]: Exactly, it was just to improve codecov.

File: test/runtests.jl (hunk 23 above)
Comment (albertomercurio) [reply test/runtests.jl:6]: We need it to increase the coverage

File: src/qobj/quantum_object_base.jl (hunk 24 above)
Comment (albertomercurio) [reply src/qobj/quantum_object_base.jl:27]: Or maybe we can add N,M. Meaning the input and output dimension. But maybe it is better to discuss about it after this PR
